### PR TITLE
 Bug 1162180 - [accessibility] add proper roles and aria attributes automatically

### PR DESCRIPTION
**Pull request by gasolin**

### CODE REVIEW

File: gaia-switch.js [hunk 1]
```diff
@@ -133,8 +146,13 @@ module.exports = component.register('gaia-switch', {
         this.els.handle.style.transform = '';
         this.els.handle.style.transition = '';
 
-        if (value) { this.setAttr('checked', ''); }
-        else { this.removeAttr('checked'); }
+        if (value) {
+          this.setAttr('checked', '');
```
Comment (gasolin) [inline gaia-switch.js]: seems need to use `setAttr` here, or the animation will be broken.

Comment (wilsonpage) [inline gaia-switch.js]: `.setAttr()` is a special version of `setAttribute()` provided by `gaia-component.js` base-class. It sets attrs inside shadow-root and on `:host`.

File: gaia-switch.js [hunk 2]
```diff
@@ -44,11 +44,23 @@ module.exports = component.register('gaia-switch', {
     this.disabled = this.getAttribute('disabled');
     this.checked = this.getAttribute('checked');
 
+    // offload the initiate operation to save loadtime
+    setTimeout(() => {
+      this.activateTransitions();
```
Comment (wilsonpage) [inline gaia-switch.js]: Keep comment explaining what this line does.

File: gaia-switch.js [hunk 3]
```diff
@@ -84,6 +96,7 @@ module.exports = component.register('gaia-switch', {
     this.drag.on('snapped', (e) => this.onSnapped(e));
   },
 
+  // enable transitions after creation
```
Comment (wilsonpage) [inline gaia-switch.js]: Move this to `createdCallback`

**Comment by gasolin** [comment]
in gaia-switch, the test (put sample 100 times) shows 150ms improvement (3614ms -> 3458ms) when we put makeAssisible into setTimeout

File: gaia-switch.js [hunk 4]
```diff
@@ -44,11 +44,31 @@ module.exports = component.register('gaia-switch', {
     this.disabled = this.getAttribute('disabled');
     this.checked = this.getAttribute('checked');
 
+    // put non-critical operations in post-load time frame
+    // to save loadtime.
```
Comment (wilsonpage) [inline gaia-switch.js]: Please comment to say *why* this is faster.

Comment (gasolin) [inline gaia-switch.js]: According to research there's `50ms` post-load idle time when user see the web page is rendered. We could save overall loadtime by using that post-load idle time (user did not start interact with UI yet)  for non-critical operations.

Comment (wilsonpage) [inline gaia-switch.js]: Does this make sense?

 'Setting element attributes during document load can be expensive. Deferring this using setTimeout seems to improve this.'

Please add more detail if need be and update the comment :)

Comment (gasolin) [inline gaia-switch.js]: Let me try again if below comment is more clear: :sweat_smile:  The idea is to: 

`process everything not effect user interaction immediately when component is created, so the UI can load faster`. 

Set element attributes is just one of these cases.

Comment (wilsonpage) [inline gaia-switch.js]: Sorry I'm still struggling to follow. I'd like to know *why* these actions are worth deferring. Do you mean:

`process everything that doesn't affect user interaction after the component is created`?

Comment (gasolin) [inline gaia-switch.js]: Yes. 

And for a11y related actions, they do affect a part of impairable user's interaction. But as above mentioned, the `50ms`+ post-load idle time can also be utilized to serve that purpose.
